### PR TITLE
perf(gatsby-plugin-contentful): drop the super expensive fixids cycle

### PR DESCRIPTION
**Pull request by pvdz**
This is the diff @axe312ger put up in https://github.com/gatsbyjs/gatsby/pull/25948 but with the last necessary fix.

Also leaving some comments to double check.

This change drops the `fixids` function call which was applied to the entire model and is very expensive to the Contentful data set. It was also no longer needed so this is a huge win for sites sourcing from Contentful.

### CODE REVIEW

File: packages/gatsby-source-contentful/src/__tests__/__snapshots__/gatsby-node.js.snap
Comment (pvdz) [inline packages/gatsby-source-contentful/src/__tests__/__snapshots__/gatsby-node.js.snap:4]: Changes in this file should basically only be dropping a `c` prefix from all ids

File: packages/gatsby-source-contentful/src/normalize.js
Comment (pvdz) [inline packages/gatsby-source-contentful/src/normalize.js:90]: @axe312ger this was the missing fix to make the tests pass

File: packages/gatsby-source-contentful/src/normalize.js
Comment (pvdz) [inline packages/gatsby-source-contentful/src/normalize.js:106]: I also applied it to these two, although it did not make a difference in the tests. @axe312ger is this a redundant change or good to keep?

File: packages/gatsby-source-contentful/src/normalize.js
Comment (pvdz) [inline packages/gatsby-source-contentful/src/normalize.js:454]: Tried dropping this entirely but that does break some tests so let's keep it for now. I think it makes sense for there to be a `contentful_id` to recognize atomic ids for content from Contentful

**Comment by vladar** [comment]
I am curious why did we have this `fixId` in the first place and why we don't need it anymore? Was it some workaround for old inference quirks?

**Comment by pvdz** [comment]
Yeah. I'm sorry I forgot the history already but the tldr is that it was useful at some point, no longer at some point, but nobody dared removing it. It's a huge drain on the sourcing time as it has to traverse the entire model, just like inference would.

**Review by wardpeet — COMMENTED** [review]
In theory this is a breaking change if people for some reason have hard-coded ids.

File: packages/gatsby-source-contentful/src/fetch.js
Comment (pieh) [inline packages/gatsby-source-contentful/src/fetch.js:115]: Are those `sys` objects accessible via schema? If so - that's breaking change

**Review by pieh — COMMENTED** [review]
It seems like there is few of things that are (or might be) breaking changes - this would need major version bump? (not a blocker, just let's try not to break user sites that might rely on those)

**Comment by pvdz** [comment]
Since not every node will get a `contentful_id` anymore, this is a potentially breaking change anyways. I'm not sure if there's anyone who depends on this, but considering that it's possible there probably is.
I think it was already the intention to release a major for this. It'll be up to Contentful to inform people that there's a major bump (not covered by `^`). They are ultimately the code owner to this plugin.
In addition, there's another PR that will land soonish that also needs a major bump. But I suspect that one is a little further out.

File: packages/gatsby-source-contentful/src/__tests__/__snapshots__/gatsby-node.js.snap
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/__tests__/__snapshots__/gatsby-node.js.snap:4]: yes, correct :)

File: packages/gatsby-source-contentful/src/normalize.js
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/normalize.js:454]: yes, we need to be able to identify content via contentful_id  👍

**Comment by axe312ger** [comment]
> Since not every node will get a `contentful_id` anymore, this is a potentially breaking change anyways. I'm not sure if there's anyone who depends on this, but considering that it's possible there probably is.
> I think it was already the intention to release a major for this. It'll be up to Contentful to inform people that there's a major bump (not covered by `^`). They are ultimately the code owner to this plugin.
> In addition, there's another PR that will land soonish that also needs a major bump. But I suspect that one is a little further out.

This is correct. This is a potential breaking change has to be a major bump. The original idea was to release this one together with #25249, but the Rich Text fix will need more time to ready for release and will introduce some refactoring work for everyone using Rich Text.


Whats the usual strategy to inform the community about major bumps? I guess the Gatsby newsletter and blog?

I can make sure that Contentful will use their channels to inform people :)

File: packages/gatsby-source-contentful/src/fetch.js
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/fetch.js:115]: contentful_id will still available, see discussion

**Comment by pvdz** [comment]
A major bump by itself is not a big deal, especially not here, because the most common way of depending on the package is by `^`, which would not bump a major. So worst case, your customers are not getting the new updates until they realize they have to bump explicitly.

Ok, let's prepare a major bump. @axe312ger I assume there's no objections on your side since you haven't left any :)

File: packages/gatsby-source-contentful/src/normalize.js
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/normalize.js:90]: Hmm, I wonder how the nodes can be of type [Link](https://www.contentful.com/developers/docs/concepts/links/) while `entryList` should contain a fully resolved Contentful API response:

https://github.com/gatsbyjs/gatsby/blob/contentful_drop_fixids/packages/gatsby-source-contentful/src/gatsby-node.js#L169-L186

File: packages/gatsby-source-contentful/src/normalize.js
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/normalize.js:106]: See comment above. We should not have links in here at the first place.

It might be that the test fixtures are wrong/outdated.

**Review by axe312ger — CHANGES_REQUESTED** [review]

**Comment by pvdz** [comment]
@axe312ger was able to resolve outstanding issues. Provided tests hold this PR is good to go. It will require a major bump so let's make sure the current minor branch is stable before we do merge.

**Review by axe312ger — APPROVED** [review]

**Comment by pieh** [comment]
Merge commit message proposal:

```
perf(gatsby-plugin-contentful): drop the super expensive fixids cycle

BREAKING CHANGE: 
If you were relying on the `contentful_id` on the `sys` object in your queries  those are no longer exposed and you can safely change them to the `id` property on that `sys` object. 

If you were relying on the `id` property in the `sys` object, you should be aware that it is no longer "normalized". In particular, it will no longer get a 'c' prefixed to the id.
```

**Comment by axe312ger** [comment]
We published a new canary version. Feel free to give it a try:

`npm i gatsby-source-contentful@next`

It does contain the rich text fixes from https://github.com/gatsbyjs/gatsby/pull/25249

**Comment by pvdz** [comment]
Published in `gatsby-source-contentful@3.0.0` (together with `gatsby@2.24.82`, although that shouldn't matter so much)

**Pull request merged**
